### PR TITLE
More STJ updates for .NET 9

### LINKED ISSUE #42592: Create/Update conceptual docs for new System.Text.Json features in .NET 9
### Describe the issue or suggestion

Creating an issue to track creating new STJ conceptual docs in .NET 9, here's a list of key new features:

- [x] Document [JsonSchemaExporter](https://github.com/dotnet/runtime/issues/102788).
- [x] Document [multiple root-level value support](https://github.com/dotnet/runtime/issues/33030#issuecomment-2197778133).
- [ ] Document the newly added [`RespectNullableAnnotations`](https://github.com/dotnet/runtime/issues/100144#issuecomment-2123233306) and [`RespectRequiredConstructorParameters`](https://github.com/dotnet/runtime/issues/100075#issuecomment-2073231418) flags, including caveats and how the two interact.
- [x] Add [`JsonEnumMemberAttribute` examples](https://github.com/dotnet/runtime/issues/74385) when using `JsonStringEnumConverter`.
- [x] Document [out-of-order metadata read support](https://github.com/dotnet/runtime/issues/72604#issuecomment-1904569859).
- [x] Document [new indentation customization options](https://github.com/dotnet/runtime/issues/63882#issuecomment-1852620212).
- [x] Document [support for property order manipulation in `JsonObject`](https://github.com/dotnet/runtime/issues/102932).
- [ ] Update contract customization document to include the [newly exposed metadata APIs](https://github.com/dotnet/runtime/issues/102078), where appropriate.
- [x] Include `JsonElement.DeepEquals` blurb in [the JsonDocument guide](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-dom#use-jsondocument).
- [x] Include reference to the new [`JsonSerializerOptions.Web` singleton](https://github.com/dotnet/runtime/pull/94370) where appropriate.

I plan on drafting a new blog post that should hopefully feed into most of the content listed above.

cc @gewarren

---
[Associated WorkItem - 320283](https://dev.azure.com/msft-skilling/Content/_workitems/edit/320283)

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #42592.

- Nullable annotation support
- Non-optional constructor parameter support
- Custom enum member names

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/whats-new/dotnet-9/libraries.md](https://github.com/dotnet/docs/blob/5c3f57f5919d20b82a34df239963482c60cb0dbf/docs/core/whats-new/dotnet-9/libraries.md) | [docs/core/whats-new/dotnet-9/libraries](https://review.learn.microsoft.com/en-us/dotnet/core/whats-new/dotnet-9/libraries?branch=pr-en-us-43168) |
| [docs/standard/serialization/system-text-json/customize-properties.md](https://github.com/dotnet/docs/blob/5c3f57f5919d20b82a34df239963482c60cb0dbf/docs/standard/serialization/system-text-json/customize-properties.md) | [docs/standard/serialization/system-text-json/customize-properties](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/customize-properties?branch=pr-en-us-43168) |
| [docs/standard/serialization/system-text-json/nullable-annotations.md](https://github.com/dotnet/docs/blob/5c3f57f5919d20b82a34df239963482c60cb0dbf/docs/standard/serialization/system-text-json/nullable-annotations.md) | [docs/standard/serialization/system-text-json/nullable-annotations](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/nullable-annotations?branch=pr-en-us-43168) |
| [docs/standard/serialization/system-text-json/required-properties.md](https://github.com/dotnet/docs/blob/5c3f57f5919d20b82a34df239963482c60cb0dbf/docs/standard/serialization/system-text-json/required-properties.md) | [docs/standard/serialization/system-text-json/required-properties](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/required-properties?branch=pr-en-us-43168) |
| [docs/standard/serialization/system-text-json/source-generation.md](https://github.com/dotnet/docs/blob/5c3f57f5919d20b82a34df239963482c60cb0dbf/docs/standard/serialization/system-text-json/source-generation.md) | [docs/standard/serialization/system-text-json/source-generation](https://review.learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/source-generation?branch=pr-en-us-43168) |

### CODE REVIEW

File: docs/standard/serialization/system-text-json/nullable-annotations.md [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+:::code language="csharp" source="snippets/nullable-annotations/Nullable.cs" id="Serialization":::
+
+Similarly, <xref:System.Text.Json.JsonSerializerOptions.RespectNullableAnnotations> enforces nullability on deserialization. The following code snippet throws a <xref:System.Text.Json.JsonException> during serialization with a message like:
+
+> The constructor parameter 'Name' on type 'Person' doesn't allow null values. Consider updating its nullability annotation.
+
+:::code language="csharp" source="snippets/nullable-annotations/Nullable.cs" id="Deserialization":::
+
+> [!TIP]
```
Comment (eiriktsarpalis) [inline docs/standard/serialization/system-text-json/nullable-annotations.md:23]: I neglected to put this in the blog, but would it be possible to add passing mention that the feature also recognizes the `[NotNull]`, `[AllowNull]`, `[MaybeNull]` and `[DisallowNull]` attributes? They're attributes used by the compiler to fine-tune annotations in getters and setters. I don't think we need examples, but I have a few tests highlighting the use case for your reference:

https://github.com/dotnet/runtime/blob/e16be4174212fd2abdfce2cfefb613f886264a54/src/libraries/System.Text.Json/tests/Common/NullableAnnotationsTests.cs#L560-L608

**Review by eiriktsarpalis — APPROVED** [review]
Thanks!

File: docs/standard/serialization/system-text-json/nullable-annotations.md (hunk 1 above)
Comment (BillWagner) [reply docs/standard/serialization/system-text-json/nullable-annotations.md:23]: If you choose to mention these attributes, you can link to [this](https://learn.microsoft.com/dotnet/csharp/language-reference/attributes/nullable-analysis) for the full docs on the attributes.

File: docs/standard/serialization/system-text-json/nullable-annotations.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+## Limitations
+
+Due to how non-nullable reference types are implemented, this feature comes with some important limitations. Familiarize yourself with these limitations before turning the feature on. The root of the issue is that reference type nullability has no first-class representation in intermediate language (IL). As such, the expressions `MyPoco` and `MyPoco?` are indistinguishable from the perspective of run-time reflection. While the compiler will try to make up for that by [emitting attribute metadata where possible](https://sharplab.io/#v2:D4AQTAjAsAULBOBTAxge3gEwAQFkCeACqmgBQgQAMWAcqgC7UCuANswMp3wCWAdgOYAaLOQoB+Gi2YBDAEbNEHbvwCUAbiA=), this metadata is restricted to non-generic member annotations that are scoped to a particular type definition. This is the reason that the flag only validates nullability annotations that are present on non-generic properties, fields, and constructor parameters. System.Text.Json does not support nullability enforcement on:
```
[the remark this answers is not in the document]
Comment (BillWagner) [reply docs/standard/serialization/system-text-json/nullable-annotations.md:28]: I might add text in the link that notes where this link goes:

```suggestion
Due to how non-nullable reference types are implemented, this feature comes with some important limitations. Familiarize yourself with these limitations before turning the feature on. The root of the issue is that reference type nullability has no first-class representation in intermediate language (IL). As such, the expressions `MyPoco` and `MyPoco?` are indistinguishable from the perspective of run-time reflection. While the compiler will try to make up for that by [emitting attribute metadata where possible (sharplab)](https://sharplab.io/#v2:D4AQTAjAsAULBOBTAxge3gEwAQFkCeACqmgBQgQAMWAcqgC7UCuANswMp3wCWAdgOYAaLOQoB+Gi2YBDAEbNEHbvwCUAbiA=), this metadata is restricted to non-generic member annotations that are scoped to a particular type definition. This is the reason that the flag only validates nullability annotations that are present on non-generic properties, fields, and constructor parameters. System.Text.Json does not support nullability enforcement on:
```

**Review by BillWagner — APPROVED** [review]
This LGTM @gewarren 

Let's :shipit:

**Pull request merged**